// express/index.js
import { createServer } from "node:http";
import { createReadStream, existsSync, statSync } from "node:fs";
import path from "node:path";

const MIME = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".svg": "image/svg+xml"
};

function matchPath(pattern, pathname) {
  if (pattern === "*" || pattern === pathname) return { ok: true, params: {} };
  const names = [];
  const regex = pattern
    .split("/")
    .map((part) => {
      if (!part.startsWith(":")) return part.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      names.push(part.slice(1));
      return "([^/]+)";
    })
    .join("/");
  const found = pathname.match(new RegExp(`^${regex}$`));
  if (!found) return { ok: false, params: {} };
  return { ok: true, params: Object.fromEntries(names.map((name, index) => [name, decodeURIComponent(found[index + 1])])) };
}

function decorateResponse(res) {
  res.status = (code) => {
    res.statusCode = code;
    return res;
  };
  res.json = (data) => {
    res.setHeader("content-type", "application/json; charset=utf-8");
    res.end(JSON.stringify(data));
  };
  res.send = (data) => {
    if (typeof data === "object") return res.json(data);
    res.end(data);
  };
  res.sendFile = (filePath) => {
    res.setHeader("content-type", MIME[path.extname(filePath)] || "application/octet-stream");
    createReadStream(filePath).pipe(res);
  };
  return res;
}

async function parseJson(req) {
  if (!["POST", "PUT", "PATCH", "DELETE"].includes(req.method)) return;
  const chunks = [];
  for await (const chunk of req) chunks.push(chunk);
  const body = Buffer.concat(chunks).toString("utf8");
  req.body = body ? JSON.parse(body) : {};
}

export default function express() {
  const routes = [];
  const middlewares = [];
  return {
    use(handler) {
      middlewares.push(handler);
    },
    get(route, handler) {
      routes.push({ method: "GET", route, handler });
    },
    post(route, handler) {
      routes.push({ method: "POST", route, handler });
    },
    delete(route, handler) {
      routes.push({ method: "DELETE", route, handler });
    },
    listen(port, callback) {
      const server = createServer(async (req, res) => {
        decorateResponse(res);
        const url = new URL(req.url, `http://${req.headers.host}`);
        req.path = url.pathname;
        req.query = Object.fromEntries(url.searchParams.entries());
        try {
          await parseJson(req);
          for (const middleware of middlewares) {
            const handled = await middleware(req, res);
            if (handled || res.writableEnded) return;
          }
          for (const route of routes) {
            if (route.method !== req.method) continue;
            const match = matchPath(route.route, req.path);
            if (!match.ok) continue;
            req.params = match.params;
            return route.handler(req, res);
          }
          res.status(404).json({ error: "Not found" });
        } catch (error) {
          res.status(500).json({ error: error.message });
        }
      });
      return server.listen(port, callback);
    }
  };
}

express.json = () => async () => false;

express.static = (root) => async (req, res) => {
  if (req.method !== "GET" && req.method !== "HEAD") return false;
  const requested = req.path === "/" ? "/index.html" : req.path;
  const filePath = path.resolve(root, `.${decodeURIComponent(requested)}`);
  if (!filePath.startsWith(path.resolve(root))) return false;
  if (!existsSync(filePath) || statSync(filePath).isDirectory()) return false;
  res.setHeader("content-type", MIME[path.extname(filePath)] || "application/octet-stream");
  createReadStream(filePath).pipe(res);
  return true;
};
